src/awkward/operations/ak_nan_to_num.py

```diff
@@ -54,82 +56,47 @@ def nan_to_num(
 
 def _impl(array, copy, nan, posinf, neginf, highlevel, behavior):
     behavior = behavior_of(array, behavior=behavior)
-
-    broadcasting_ids = {}
-    broadcasting = []
-
-    layout = ak.operations.to_layout(array)
-    broadcasting.append(layout)
-
-    nan_layout = ak.operations.to_layout(nan, allow_other=True)
-    if isinstance(nan_layout, ak.contents.Content):
-        broadcasting_ids[id(nan)] = len(broadcasting)
-        broadcasting.append(nan_layout)
-
-    posinf_layout = ak.operations.to_layout(posinf, allow_other=True)
-    if isinstance(posinf_layout, ak.contents.Content):
-        broadcasting_ids[id(posinf)] = len(broadcasting)
-        broadcasting.append(posinf_layout)
-
-    neginf_layout = ak.operations.to_layout(neginf, allow_other=True)
-    if isinstance(neginf_layout, ak.contents.Content):
-        broadcasting_ids[id(neginf)] = len(broadcasting)
-        broadcasting.append(neginf_layout)
-
-    if len(broadcasting) == 1:
-
-        def action(layout, backend, **kwargs):
-            if isinstance(layout, ak.contents.NumpyArray):
-                return ak.contents.NumpyArray(
+    assert array is not None
+    assert nan is not None
+
+    inputs = []
+    input_is_scalar = []
+    for obj in (array, nan, posinf, neginf):
+        layout, layout_is_scalar = _to_layout_detailed(
+            unset if obj is None else obj,
+            allow_record=True,
+            allow_other=True,
+            regulararray=True,
+        )
+        inputs.append(layout)
+        input_is_scalar.append(layout_is_scalar)
+
+    def action(inputs, backend, **kwargs):
+        layout, nan_layout, posinf_layout, neginf_layout = inputs
+        if all(
+            x.purelist_depth == 1 for x in inputs if isinstance(x, ak.contents.Content)
+        ):
+            return (
+                ak.contents.NumpyArray(
                     backend.nplike.nan_to_num(
                         layout.data,
-                        nan=nan,
-                        posinf=posinf,
-                        neginf=neginf,
+                        nan=nan_layout.data,
+                        posinf=None if posinf_layout is unset else posinf_layout.data,
+                        neginf=None if neginf_layout is unset else neginf_layout.data,
```

Comment:
This preserves some of the performance aspect to not requiring a broadcast (the default case, but not the all-scalars case). Functionally we could just remove this along with the above `unset if obj is None` logic, but this ensures we only create an array if an argument is given. The most general solution would explicitly check for scalars, and only broadcast those which aren't scalar (as originally written).